pennylane/transforms/decompositions/clifford_t/clifford_t_transform.py

```diff
@@ -0,0 +1,479 @@
+
+    while len(copied_ops) > 0:
+        curr_gate = copied_ops.pop(0)
+
+        # If gate is not to be merged, let it go
+        if merge_ops is not None and curr_gate.name not in merge_ops:
+            merged_ops.append(curr_gate)
+            continue
+
+        # If gate is not to be merged, update counter
+        if curr_gate.name in merge_ops:
+            number_ops += 1
+
+        # Find the next gate that acts on the same wires
+        next_gate_idx = find_next_gate(curr_gate.wires, copied_ops)
+        if next_gate_idx is None:
+            merged_ops.append(curr_gate)
+            continue
+
+        # Initialize the current angle and iterate until end of merge
+        cumulative_angles = qml.math.array(curr_gate.parameters, dtype=float)
+        next_gate = copied_ops[next_gate_idx]
+        while curr_gate.name == next_gate.name and curr_gate.wires == next_gate.wires:
+            cumulative_angles = cumulative_angles + qml.math.array(next_gate.parameters)
+            # Check if the subsequent gate exists in the vicinity
+            copied_ops.pop(next_gate_idx)
+            next_gate_idx = find_next_gate(curr_gate.wires, copied_ops)
+            if next_gate_idx is None:
+                break
+            next_gate = copied_ops[next_gate_idx]
+
+        # Replace the current gate, add it to merged list and pop it from orginal one
+        merged_ops.append(curr_gate.__class__(*cumulative_angles, wires=curr_gate.wires))
+
+    return merged_ops, number_ops
+
+
+# pylint: disable= too-many-nested-blocks, too-many-branches, too-many-statements, unnecessary-lambda-assignment
+@transform
+def clifford_t_decomposition(
+    tape: QuantumTape,
+    epsilon=1e-4,
+    max_expansion=6,
+    method="sk",
+    **method_kwargs,
+) -> (Sequence[QuantumTape], Callable):
+    r"""Unrolls a circuit into the Clifford+T basis using the optimal ancilla-free approximation of :class:`~.RZ` operations.
+
+    This method first decomposes the gate operations to a basis comprised of Clifford, :class:`~.T`, :class:`~.RZ` and
+    :class:`~.GlobalPhase` operations (and their adjoints). The Clifford gates include the following PennyLane operations:
+
+    - Single qubit gates - :class:`~.Identity`, :class:`~.PauliX`, :class:`~.PauliY`, :class:`~.PauliZ`,
+      :class:`~.SX`, :class:`~.S`, and :class:`~.Hadamard`.
+    - Two qubit gates - :class:`~.CNOT`, :class:`~.CY`, :class:`~.CZ`, :class:`~.SWAP`, and :class:`~.ISWAP`.
+
+    Then the leftover single qubit :class:`~.RZ` operations are approximated in the Clifford+T basis with
+    :math:`\epsilon > 0` error using the Solovay-Kitaev algorithm described in
+    `Dawson and Nielsen (2005) <https://arxiv.org/abs/quant-ph/0505030>`_.
+
+    Args:
+        tape (QuantumTape): A quantum tape
+        epsilon (float): The maximum permissible error. Defaults to ``0.0001``.
+        max_expansion (int): The depth to be used for tape expansion before manual decomposition to Clifford+T basis is applied.
```

Comment:
What is the motivation for exposing this here?

Reply:
The motivation is to allow users to be able to use the simplified decomposition routine (`_rot_decompose`) on the original (or not fully expanded) circuit as this can result in better (shorter) decomposition than the standard expansion workflow.

Reply:
it's also passed directly to `qml.compile` internally, and that transform exposes this parameter as well.